smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSource.java

```diff
@@ -111,52 +124,59 @@ public KafkaSource(Vertx vertx,
-                                .with(
-                                        a -> log.executedConsumerRevokedRebalanceListener(group),
-                                        t -> log.unableToExecuteConsumerRevokedRebalanceListener(group, t));
-                    });
                 });
+
+        if (rebalanceListener.isPresent()) {
+            KafkaConsumerRebalanceListener listener = rebalanceListener.get();
+            // If the re-balance assign fails we must resume the consumer in order to force a consumer group
+            // re-balance. To do so we must wait until after the poll interval time or
+            // poll interval time + session timeout if group instance id is not null.
+            // We will retry the re-balance consumer listener on failure using an exponential backoff until
+            // we can allow the kafka consumer to do it on its own. We do this because by default it would take
+            // 5 minutes for kafka to do this which is too long. With defaults consumerReEnableWaitTime would be
+            // 500000 millis. We also can't simply retry indefinitely because once the consumer has been paused
+            // for consumerReEnableWaitTime kafka will force a re-balance once resumed.
+            final long consumerReEnableWaitTime = Long.parseLong(
+                    kafkaConfiguration.getOrDefault(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "300000"))
+                    + (kafkaConfiguration.get(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG) == null ? 0L
+                            : Long.parseLong(
+                                    kafkaConfiguration.getOrDefault(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG,
+                                            "10000")))
+                    + 11_000L; // it's possible that it might expire 10 seconds before when we need it to
+
+            kafkaConsumer.partitionsAssignedHandler(set -> {
+                kafkaConsumer.pause();
+                log.executingConsumerAssignedRebalanceListener(group);
+                listener.onPartitionsAssigned(kafkaConsumer, set)
+                        .onFailure().invoke(t -> log.unableToExecuteConsumerAssignedRebalanceListener(group, t))
+                        .onFailure().retry().withBackOff(Duration.ofSeconds(1), Duration.ofSeconds(10))
+                        .expireIn(consumerReEnableWaitTime)
+                        .subscribe()
+                        .with(
+                                a -> {
+                                    log.executedConsumerAssignedRebalanceListener(group);
+                                    commitHandler.partitionsAssigned(vertx.getOrCreateContext(), set);
+                                    kafkaConsumer.resume();
+                                },
+                                t -> {
+                                    log.reEnablingConsumerforGroup(group);
+                                    commitHandler.partitionsAssigned(vertx.getOrCreateContext(), set);
```

Comment:
Line 158 and 163 are the only actual changes. We have to inform the commit handler that a re-balance took place so it can reset its internal data structurers.